---
title: |
  | \textbf{Supplementary Figures for:}
  | Genomic Diversity of SARS-CoV-2 During Early Introduction into the United Sates National Capital Region
date: "`r format(Sys.time(), '%B %d, %Y')`"
always_allow_html: true
output:
  bookdown::pdf_book:
    base_format: rticles::elsevier_article
geometry: "left=4cm, right=4cm, top=4cm, bottom=4cm"
header-includes:
  \usepackage[dvipsnames]{xcolor}
  \usepackage{geometry}
  \definecolor{gray80}{rgb}{0.8,0.8,0.8}
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir="../")
knitr::opts_chunk$set(fig.width=7.5, fig.height=6, fig.path='../figures/', warning=FALSE, message=FALSE,cache=FALSE,dev=c('png','pdf'))
```

```{r, load-libraries, echo=FALSE, message=FALSE}
library(ggplot2)
library(pheatmap)
library(ape)
library(stringr)
library(dplyr)
library(seqinr)
source("src/plot_funcs_supp.R") # Supplementary figure plotting functions

```

\newpage
\newgeometry{left=2cm,right=2cm,top=2cm,bottom=2cm}

```{r load-metadata, echo=FALSE}

## load metadata that will be used throughout
seq_metadata <- read.csv("data/sample_metadata.csv",header=TRUE)

# ensure dates are in date format
seq_metadata$collection_date <- as.Date(seq_metadata$collection_date)

# ensure numeric columns are numeric

seq_metadata <- seq_metadata %>%
  mutate(age=replace(age,age=="unknown",NA)) %>%
  mutate(age=as.numeric(age)) %>%
  mutate(diagnostic_ct_value=replace(diagnostic_ct_value,diagnostic_ct_value=="unknown",NA)) %>%
  mutate(diagnostic_ct_value=as.numeric(diagnostic_ct_value)) %>%
  mutate(days_from_onset=replace(days_from_onset,days_from_onset=="unknown",NA)) %>%
  mutate(days_from_onset=as.numeric(days_from_onset))

```

```{r load-heatmap-data, echo=FALSE}

# load clinical metdata
md_meta <- seq_metadata
md_meta <- md_meta %>% filter(genome=="Yes")

# read in alignment of all sequences
align <- read.alignment("data/all_submitted_jhu_sequences.fasta","fasta")
align <- as.matrix(align)
rownames(align) <- substr(rownames(align),13,22)
rownames(align) <- str_replace(rownames(align),"..-HP","MDHP-")
align <- subset(align,rownames(align)!="MDHP-00146") # sequence that did not produce a complete genome
  
# load reference genome
ref <- read.fasta("data/reference.fasta")
ref <- as.vector(ref$MN908947.3)

# function to determine which positions have all N or only ref and N
get.novar.cols <- function(pos,aln,ref){
    aln_col <- aln[,pos]
    aln_col <- as.vector(aln_col)
    col_vals <- unique(aln_col)
    if (length(col_vals)==1 && col_vals=="n") { return(FALSE) }
    if (length(col_vals)==2 && ("n" %in% col_vals) && (ref[pos] %in% col_vals)) { return(FALSE) }
    return(TRUE)
}
  
cols_to_keep <- sapply(1:29903,get.novar.cols,align,ref)

# function to change variant values to numbers
vars.to.numeric <- function(pos,aln,ref){
  aln_col <- aln[,pos]
  aln_col <- as.vector(aln_col)
  sapply(aln_col,function(x) ifelse(x==ref[pos],0,ifelse(x %in% c("a","c","t","g"),1,2)),USE.NAMES=FALSE)
}

# change variant values to numbers
mut_num <- sapply(1:29903,vars.to.numeric,align,ref)
rownames(mut_num) <- rownames(align)
colnames(mut_num) <- seq(1,29903)

# remove columns with all "N" or only "N" and reference
mut_mat <- mut_num[,cols_to_keep]

# keep only variants that show up in at least 3 samples
cols_with_vars <- as.vector(apply(mut_mat,2,function(x) sum(x)>=3))
mut_mat <- mut_mat[,cols_with_vars]

# select which annotations to keep
annots <- md_meta %>% select("name","nextstrain_clade","nextstrain_jhhs_clade","pangolin_clade")

# ensure data and annotations are in the same order
mut_mat <- mut_mat[annots$name,]

```

# Supplemental Figure 1. Geographic distribution of positive diagnoses
```{r, map-dx, echo=FALSE, warning=FALSE}
#-------------------------------------------------
# load libraries
library(maptools)
library(rgdal)
library(sf)
library(lwgeom)
library(mapplots)
library(RColorBrewer)
library(dplyr)

#-------------------------------------------------
# set up aesthetics
water_col <- "lightblue"
zip3_fill_col <- rev(RColorBrewer::brewer.pal(n = 4, name = "Greys"))[2:4]
zip3_border_col <- "gray90"
zip3_border_width <- 0.5
state_border_col <- "gray90"
state_border_width <- 0.1
maryland_border_col <- "gray30"
maryland_border_width <- 0.1
county_border_col <- "gray90"
county_border_width <- 0.1
baltimore_border_col <- "gray90"
baltimore_border_width <- 0.25
water_line_width <- 1

#-------------------------------------------------
# set up viewbox
asp <- 1
center <- c(-77.25, 38.85)
span <- c(3.8, 2.1)
new_center <- c(0, 0)
longlat_spacing <- 1
countries <- c("United States")
regions <- ""
subregions <- ""
states <- c("Maryland", "Delaware", "Virginia", "West Virginia", "Pennsylvania", "New Jersey")
states_fips <- c(24, 10, 51, 54, 42, 34)
xlim <- c(-span[1]/2+new_center[1], span[1]/2+new_center[1])*1e5
ylim <- c(-span[2]/2+new_center[2], span[2]/2+new_center[2])*1e5

rotate.sphere.old <- function(long, lat, a, axis=c(0,1,0)) {
  
  # rotates a long/lat pair 'a' degrees about an axis (defaults to y-axis)
  # z-axis is north/south pole of earth
  # x-axis comes out of the earth at (0, 0)
  # y-axis comes out of the earth at (0, 90)
  # returns the rotated longitude and latitude:
  #   where the new 'north pole' is where the z-axis comes out of the rotated sphere
  
  # define constants
  pi <- 3.14159265358979
  
  # convert lat, long, 'a' to radians
  az <- long * pi/180
  incl <- (90-lat) * pi/180
  a <- a * pi/180
  
  # convert spherical coordinates to cartesian coordinates
  x <- cos(az) * sin(incl)
  y <- sin(az) * sin(incl)
  z <- cos(incl)
  
  # rotate 'a' degrees about rotation axis
  c1 <- axis[1]
  c2 <- axis[2]
  c3 <- axis[3]
  x_p <- x*cos(a) + (1-cos(a))*(c1*c1*x + c1*c2*y + c1*c3*z) + (c2*z - c3*y)*sin(a)
  y_p <- y*cos(a) + (1-cos(a))*(c2*c1*x + c2*c2*y + c2*c3*z) + (c3*x - c1*z)*sin(a)
  z_p <- z*cos(a) + (1-cos(a))*(c3*c1*x + c3*c2*y + c3*c3*z) + (c1*y - c2*x)*sin(a)
  
  # convert back to spherical coordinates
  az_p <- atan2(y_p, x_p)
  incl_p <- acos(z_p)
  
  # convert azimuth/inclination to longitude/latitude
  long_p <- az_p * 180/pi
  lat_p <- 90 - incl_p * 180/pi
  
  return(c(long_p, lat_p))
}


# Figure 1D
zip3_dx <- read.csv("data/zip3_dx_counts.csv")

zip3.tr <- sf::st_read('data/shapefiles/zip3_mdhp.shp', quiet=TRUE)
county.tr <- sf::st_read('data/shapefiles/county_mdhp.shp', quiet=TRUE)
state.tr <- sf::st_read('data/shapefiles/state_mdhp.shp', quiet=TRUE)
country.tr <- sf::st_read('data/shapefiles/country_mdhp.shp', quiet=TRUE)
water.tr <- sf::st_read('data/shapefiles/water_mdhp.shp', quiet=TRUE)
CRSstring <- paste("+proj=lcc +lat_1=33 +lat_2=45 +lat_0=", (center[2]-0.005), " +lon_0=", center[1], " +x_0=0 +y_0=0 +ellps=GRS80 +datum=WGS84 +units=m +no_defs", sep="")
water.tr <- sf::st_transform(water.tr, CRS(CRSstring))

#-------------------------------------------------
# create object for fill for Chesapeake + Delaware Bays
c_coord <- st_coordinates(country.tr)
bay_water <- c_coord[4100:8377,]

#-------------------------------------------------
# set up zip data

# link totals with zip3 shapefile
count.mat <- function(zip) {
  ret <- 0
  zip_present <- zip3_dx$zip3 == zip
  if(sum(zip_present) > 0) {
    ret <- as.numeric(zip3_dx$count[zip_present])
  }
  return(ret)
}
count <- sapply(as.matrix(zip3.tr$ZCTA3CE), count.mat)

# calculate color ramp from 0 to 1 for all places in 'place'
ramp <- (1 - count/max(zip3_dx$count))
ramp[ramp < 0] <- 0

# convert ramp to rgb colors in hex string
col_ramp <- colorRamp(zip3_fill_col)
zip_col <- rgb(t(sapply(ramp, col_ramp)/255))
zip_col[ramp==1] <- "#FFFFFF"

#-------------------------------------------------
# create empty plot
plot(0,0, type='n', xlim=xlim, ylim=ylim, asp=asp, axes=F, xlab="", ylab="")
# add water fill rectangle as base
rect(-360*1e6, -180*1e6, 360*1e6, 180*1e6, col=water_col, border=NA)
# plot zip3 fill first
plot(zip3.tr, xlim=xlim, ylim=ylim, asp=asp,
     col=zip_col, border=NA, add=TRUE)
# overlay bay water
polypath(bay_water,
         col=water_col, border=FALSE)
# add other lakes / water polygons
plot(water.tr$geometry, xlim=xlim, ylim=ylim, asp=asp,
     col=water_col, border=NA, add=TRUE)
# add rivers / water lines
# lines(waterlines.tr, xlim=xlim, ylim=ylim, asp=asp,
#       col=water_col, lwd=water_line_width)
# add county borders
plot(county.tr$geometry[county.tr$NAME != "Baltimore",], xlim=xlim, ylim=ylim, asp=asp,
     col=NA, border=county_border_col, lwd=county_border_width, add=TRUE)
# add Baltimore city border
plot(county.tr$geometry[county.tr$NAME == "Baltimore",], xlim=xlim, ylim=ylim, asp=asp,
     col=NA, border=baltimore_border_col, lwd=baltimore_border_width, add=TRUE)
# add zip3 borders
plot(zip3.tr$geometry, xlim=xlim, ylim=ylim, asp=asp,
     col=NA, border=zip3_border_col, lwd=zip3_border_width,
     add=TRUE)
# add state borders
plot(state.tr$geometry[state.tr$STATEFP != 24,], xlim=xlim, ylim=ylim, asp=asp,
     col=NA, border=state_border_col, lwd=state_border_width, add=TRUE)
# add Maryland border
plot(state.tr$geometry[state.tr$STATEFP == 24,], xlim=xlim, ylim=ylim, asp=asp,
     col=NA, border=maryland_border_col, lwd=maryland_border_width, add=TRUE)

legend_ramp <- c(5, 50, 100, max(zip3_dx$count))/max(zip3_dx$count)
zip3_fill_col <- rev(RColorBrewer::brewer.pal(n = 4, name = "Greys"))[2:4]
col_ramp <- colorRamp(zip3_fill_col)
legend_col <- rev(rgb(t(sapply(legend_ramp, col_ramp))/255))
legend(x=-190000, y=5000, fill=legend_col, legend=c("1-5", 50, 100, max(zip3_dx$count)), border=NA, box.lty=0, bg="white", cex=1, title='Positive diagnoses')

#-------------------------------------------------

o <- order(names(which(count>0)))
lat <- as.numeric(as.matrix(zip3.tr$INTPTLAT[count>0][o]))
lon <- as.numeric(as.matrix(zip3.tr$INTPTLON[count>0][o])) - center[1]

rotated <- matrix(rotate.sphere.old(lon, lat, center[2]), ncol=2)

zip_sel <- zip3_dx$zip3 %in% zip3.tr$ZCTA3CE
zip_labels <- zip3.tr$ZCTA3CE[count>0][o]

plot_totals <- zip3_dx$count[zip_sel]
cex <- rep(0.4, length(plot_totals))
cex[plot_totals<=15] <- 0.3
cex[plot_totals==5] <- 0.25
offset <- rep(-0.1, length(plot_totals))
offset[plot_totals<=15] <- -0.075
offset[plot_totals==5] <- -0.05
for(i in seq_along(plot_totals)) {
  mapplots::add.pie(z=1,
                    x=rotated[i,1]*1e5,
                    y=rotated[i,2]*1e5,
                    radius=0,
                    border=NA,
                    labels=zip_labels[i],
                    label.dist=0,
                    pos=1,
                    offset=offset[i],
                    cex=0.75
  )
  # mapplots::add.pie(z=1,
  #                   x=(rotated[i,1]+0.125)*1e5,
  #                   y=rotated[i,2]*1e5,
  #                   radius=0,
  #                   border=NA,
  #                   labels=paste0("(n = ", zip3_dx$count[zip_sel][i], ")"),
  #                   label.dist=0,
  #                   pos=1,
  #                   offset=offset[i]+0.025,
  #                   cex=0.6
  # )
}

```

# Supplemental Figure 2. JHU-relevant SNPs.
```{r, figS2, echo=FALSE, warning=FALSE, fig.width=7.5, fig.height=9}

# order rows by nextstrain clade and jhhs clade
row_order <- order(annots[,"nextstrain_clade"],annots[,"nextstrain_jhhs_clade"],
                   annots[,"pangolin_clade"],decreasing=TRUE)
annots <- annots[row_order,]
rownames(annots) <- annots$name
annots <- select(annots,-name)

clade_heatmap <- pheatmap(mut_mat[row_order,],
        labels_col=colnames(mut_mat[row_order,]),
        labels_row=rownames(mut_mat[row_order,]),
        col=c("#FFFFFF", "gray30", "gray70"),
        xlab="mutation position",
        ylab="sample",
        fontsize_row=4,fontsize_col=4,
        margins=c(10,5),
        cluster_rows=FALSE,
        # clustering_distance_rows="canberra",
        clustering_distance_cols="euclidean",
        clustering_method="complete",
        angle_col=270,
        border_color = "gray80",
        legend = FALSE,
        annotation_row = annots,
        silent=TRUE
        )

clade_heatmap

```
Homoplasies visible at 11083, 13730, 27964

# Supplemental Figure 3. Illumina versus Nanopore allele frequency comparison.
```{r, figS3, echo=FALSE, warning=FALSE, fig.width=7.5, fig.height=3.5}

postfilt_file <- "data/all_postfilt_variants.txt"
postfilt_summary <- "data/all_postfilt_summaries.txt"
vars_file = "data/allVars_newclades.txt"

plots <- make_ont_ill_plots(postfile_file, postfilt_summary, vars_file)

grid.arrange(plots[[1]],plots[[3]],ncol=2)

```

# Supplemental Figure 4. Homopolymer and strand bias example.

# Supplemental Figure 5. Genetic distances matrix.

\newpage
# RMarkdown Environment
```{r, markdown-env, eval=TRUE, echo=FALSE}
# print out all packages used
sessionInfo()
```

```{r close_plots, echo=FALSE, results='hide'}
dev.off()
```
